src/core/trulens/core/utils/deprecation.py

```diff
@@ -0,0 +1,164 @@
+)
+
+
+def packages_dep_warn(module: str):
+    """Issue a deprecation warning for a backwards-compatibility modules.
+
+    This is specifically for the trulens_eval -> trulens module renaming and
+    reorganization.
+    """
+
+    warnings.warn(
+        f"The `{module}` module is deprecated. "
+        f"See {PACKAGES_MIGRATION_LINK} for instructions on migrating to `trulens.*` modules.",
+        DeprecationWarning,
+        stacklevel=3,
+    )
+
+
+def has_deprecated(obj: Union[Callable, Type]) -> bool:
+    """Check if a function or class has been deprecated."""
+
+    return has_moved(obj)
+
+
+def has_moved(obj: Union[Callable, Type]) -> bool:
+    """Check if a function or class has been moved."""
+
+    return (
+        hasattr(obj, "__doc__")
+        and obj.__doc__ is not None
+        and "has moved:\n" in obj.__doc__
+    )
+
+
+def function_moved(func: Callable, old: str, new: str):
+    """Issue a warning upon function call that has been moved to a new location.
+
+    Issues the warning only once.
+    """
+
+    assert hasattr(
+        func, "__name__"
+    ), "Cannot create dep message without function name."
+
+    message = (
+        f"Function `{func.__name__}` has moved:\n"
+        f"\tOld import: `from {old} import {func.__name__}`\n"
+        f"\tNew import: `from {new} import {func.__name__}`\n "
+    )
+
+    warned = False
+
+    @functools.wraps(func)
+    def _movedfunction(*args, **kwargs):
+        nonlocal warned
+
+        if not warned:
+            warnings.warn(message, DeprecationWarning, stacklevel=2)
+            warned = True
+
+        return func(*args, **kwargs)
+
+    _movedfunction.__doc__ = message
+
+    return _movedfunction
+
+
+def class_moved(
+    cls: Type,
+    old_location: Optional[str] = None,
+    new_location: Optional[str] = None,
+):
+    """Issue a warning upon class instantioation that has been moved to a new
+    location.
+
+    Issues the warning only once.
+    """
+
+    if new_location is None:
+        new_location = cls.__module__
+
+    message = (
+        f"Class `{cls.__name__}` has moved:\n"
+        + (
+            f"\tOld import: `from {old_location} import {cls.__name__}`\n"
+            if old_location
+            else ""
+        )
+        + f"\tNew import: `from {new_location} import {cls.__name__}`\n "
+        + f"See {PACKAGES_MIGRATION_LINK} for instructions on migrating to `trulens` modules."
+    )
+
+    warned = False
+    moved_class = cls
+
+    if issubclass(cls, Enum):
+        # Enums cannot be extended.
+        return cls
+
+    class _MovedClass(moved_class):
+        def __new__(cls, *args, **kwargs):
+            nonlocal warned
+
+            if not warned:
+                warnings.warn(message, DeprecationWarning, stacklevel=2)
+                warned = True
+
+            return moved_class.__new__(moved_class, *args, **kwargs)
+
+    _MovedClass.__doc__ = message
+
+    return _MovedClass
+
+
+def moved(
```

Comment:
If users have both trulens_eval installed already, and also installed new version trulens without uninstalling trulens_eval what would happen?

Reply:
Will try to enforce this with package constraints with this idea:
- Rename the `trulens-legacy` package to `trulens_eval`, starting with version 1.0.0 which is greater than existing trulens_eval version.
- In `trulens_eval` will now have requirements to install the required `trulens-*` packages.
- With new `trulens-*` releases, the constraints in `trulens_eval` will have to be updated to keep track (or at least occasionally).

Reply:
@sfc-gh-chu can you review this idea and check that the implementation of it in the PR is correct?